text/entities/0264-resource-and-entities.md

```diff
@@ -0,0 +1,852 @@
+  - An OOTB ["Env Variable Entity Detector"](#environment-variable-detector) will be specified and provided vs. requiring SDK wide ENV variables for resource detection.
+- *Additionally, Resource Provider would be responsible for understanding Entity lifecycle events, for Entities whose lifetimes do not match or exceed the SDK's own lifetime (e.g. browser session).*
+
+#### Entity Detector
+
+The Entity detector in the SDK is responsible for detecting possible entities that could identify the SDK (called "associated entities"). For Example, if the SDK is running in a kubernetes pod, it may provide an Entity for that pod. SDK Entity Detectors are only required to provide identifying attributes, but may provide descriptive attributes to ensure combined Resource contains similar attributes as today's SDK.
+
+An Entity Detector would have an API similar to:
+
+```rust
+trait EntityDetector
+  pub fn detect_entities(...) -> Result<Vec<Entity>, EntityDetectionError>>
+```
+
+Where `Result` is the equivalent of error channel in the language of choice (e.g. in Go this would be `entities, err := e.detectEntities()`).
+
+An Entity Detector MUST NOT provide two entities of the same entity type (e.g. two `host` or two `service` entities).
+
+#### Entity Merging and Resource
+
+The most important aspect of this design is how Entities will be merged to construct a Resource.
+
+We provide a simple algorithm for this behavior:
+
+- Construct a set of detected entities, `E`
+  - All entity detectors are sorted by priority (highest first)
+  - For each entity detector `D`, detect entities
+    - For each entity detected, `d'`
+      - If an entity `e'` exists in `E` with same entity type as `d'`, do one of the following:
+        - If the entity identiy and schema_url are the same, merge the descriptive attributes of `d'` into `e'`:
+          - For each descriptive attribute `da'` in `d'`
+            - If `da'.key` does not exist in `e'`, then add `da'` to `ei`
+            - otherwise, ignore.
+        - If the entity identity is the same, but schema_url is different: drop the new entity `d'`
+          *Note: We could offer configuration in this case*
+        - If the entity identity is different: drop the new entity `d'`.
+      - Otherwise, add the entity `d'` to set `E`
+- Construct a Resource from the set `E`.
+  - If all entities within `E` have the same `schema_url`, set the resource's
```

Comment:
What about the resource attributes? Are they also dependent on the entities set `E`?

Reply:
Yes.

You construct a Resource using this entity set.

- The `Resource.attributes` field contains *alL* identifying and descriptive attributes from the entity set.
- The `Resource.schema_url` field is set to the schema_url shared by all entities (if they all have the same one) or unset if they do not have the same schema_url.
- The `Resource.entities` field would contain the EntityRef proto, which just contains key names and the entity type for the entity set.

Reply:
What is an `EntityRef` proto?

Reply:
It's defined later in the document.  We'll be creating that and contributing it to OTLP as part of this OTEP.